### PR TITLE
Check for and correct data corruption during serial upload.

### PR DESCRIPTION
**Pull request by mrdunk**
When using your script with an inexpensive USB-UART adapter (converted Nokia CA-42) i was getting frequent data corruption.
The solution i present here mimics manual input;
  it uses ASCII 0x08 [backspace] to delete corrupt characters when they are echoed back to this script.

My initial prototype did this one character at a time but i found flushing the buffers often enough made this too slow.
The version presented here works one line at a time. After a whole line has been uploaded, the returned echo is compared to the sent data.
If the 2 differ, the whole line is deleted (using [backspace]) then re-entered.
Only when the whole line is correct is '\n' sent.

This pull request currently works well but the code style is a little rough.
Please feel free to either ask me continue working on it or use the concept yourself.

### CODE REVIEW

**Comment by kmpm** [comment]
Is there any noticeable performance impact, when there is no corrections being done?

**Comment by kmpm** [comment]
This will not be implemented. Not as is at least.

**Comment by mrdunk** [comment]
apologies for missing your comment back in November.

i've not actually used nodemcu-uploader since i wrote this so i'm afraid i can't remember what the performance impact was like with regards to the time taken on a single upload.
i /think/ the performance hit was minimal as it makes use of the echo response which means no additional traffic on the wire.

using my hardware set-up, my modifications made things far faster if data corruption was taken into account.
uploading text files a few hundred lines long i was getting less than 25% success rate meaning i had to re-try uploading multiple times.

please note the USB-UART adapter i was using failed a few months later so it is possible that it was the cause of the corruption.
note though that the USB-UART adapter did not present this issue when used for any other purpose.

so in summary,
if you have other users with data corruption problems
or if you ever need guaranteed success of an upload
the commit i presented would be one way to tackle the problem.

either way, i abandoned Lua as practical language on the ESP8266 in favor of C due to storage space issues so am unlikely to contribute to this project in the future.
thanks for making it available though. it was very useful while i was suing it.
